include/robots/type_traits.h

```diff
@@ -0,0 +1,43 @@
+#pragma once
+
+#define BOB_DEFINE_HAS_METHOD(traitName, methodName)      \
```

Comment:
I think this is a bit too clever. There aren't THAT many types I would just manually define the type traits for each robot type (this could be done with some macro boilerplate)

Reply:
Hmm I'm not sure exactly what you're suggesting. Do you mean that robots classes would each have e.g. a static ``IsTank`` member if they're of a tank type?

Reply:
I suppose we could also probably just check whether a class inherits from ``TankBase``

Reply:
I mean, as each of your robot types *has* a corresponding base class, that would definitely be one approach. What I was thinking of was:
```c++
template<typename T>
struct IsAckermann
{
    static constexpr bool value = false;
}

template<>
struct IsAckermann<RCCar>
{
    static constexpr bool value = true;
}
```
Where the specialisation could easily be added via a macro if you wanted

Reply:
Ahh I see.

I just tried an alternative approach to see if it works and it does! This code just checks whether a given class inherits from e.g. ``TankBase`` to determine whether it's a tank or not:
```c++
#pragma once

#define BOB_DEFINE_TRAIT(TYPE)                                                        \
    namespace TYPE {                                                                  \
    template<class Derived>                                                           \
    class TYPE##Base;                                                                 \
    }                                                                                 \
    template<class T>                                                                 \
    struct Is##TYPE                                                                   \
    {                                                                                 \
        static constexpr bool value = std::is_base_of<TYPE::TYPE##Base<T>, T>::value; \
    };

namespace BoBRobotics {
namespace Robots {

BOB_DEFINE_TRAIT(Ackermann)
BOB_DEFINE_TRAIT(Tank)
BOB_DEFINE_TRAIT(Omni2D)
BOB_DEFINE_TRAIT(UAV)

} // Robots
} // BoBRobotics
```

It does mean that we'd have to keep the ``AckermannBase`` class though.

Reply:
Awesome - as you have the base classes, this is probably the way to go! In this case, I'd probably just stick the ``std::is_base_of`` straight into the ``std::enable_if`` clauses where you use them and delete this header though.

Reply:
I feel the header could be generically useful though. There is a bit of boilerplate you have to do (i.e. forward-declaring your base classes or including the relevant headers) so it would be a faff if we wanted to use it elsewhere.

Atm we do have places where we're doing things like ``#ifdef ROBOT_TYPE_MECANUM`` where we could express it more elegantly with a ``Robots::IsOmni2D<RobotType>``. Would be handy for new code.

Reply:
In order to do anything with a robot object you'll need to have it's base class #included one way or other and the problem with replacing #ifdefs is that you often use them to guard code that won't compile e.g. linux-only robot types but....nonetheless this is probably generally useful 😄

Reply:
That's true, but if you're only using tank-type robots you wouldn't want to have to include headers for Ackermann robots just for the sake of making the ``std::enable_if``s work. We def should be able to at least reduce *some* ifdeffery this way...